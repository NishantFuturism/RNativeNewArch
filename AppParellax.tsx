import * as React from "react";
import { Button, View } from "react-native";
import Animated, {
  Extrapolate,
  interpolate,
  useAnimatedStyle,
} from "react-native-reanimated";
import Carousel from "react-native-reanimated-carousel";

// import SButton from "../../components/SButton";
import { ElementsText, window } from "./src/exampleExpo/src/constants";
import { withAnchorPoint } from "./src/exampleExpo/src/utils/anchor-point";
import { fruitItems } from "./src/exampleExpo/src/utils/items";

const colors = ["#fda282", "#fdba4e", "#800015"];

const PAGE_WIDTH = window.width;
const PAGE_HEIGHT = window.width * 1.2;

function AppParellax(props) {
  const [isAutoPlay, setIsAutoPlay] = React.useState(true);

  const baseOptions = {
    vertical: false,
    width: PAGE_WIDTH,
    height: PAGE_HEIGHT,
  } as const;

  const Card: React.FC<{
    index: number
    animationValue: Animated.SharedValue<number>
  }> = ({ index, animationValue }) => {
    const WIDTH = PAGE_WIDTH / 1.5;
    const HEIGHT = PAGE_HEIGHT / 1.5;
  
    const cardStyle = useAnimatedStyle(() => {
      const scale = interpolate(
        animationValue.value,
        [-0.1, 0, 1],
        [0.95, 1, 1],
        Extrapolate.CLAMP,
      );
  
      const translateX = interpolate(
        animationValue.value,
        [-1, -0.2, 0, 1],
        [0, WIDTH * 0.3, 0, 0],
      );
  
      const transform = {
        transform: [
          { scale },
          { translateX },
          { perspective: 200 },
          {
            rotateY: `${interpolate(
              animationValue.value,
              [-1, 0, 0.4, 1],
              [30, 0, -25, -25],
              Extrapolate.CLAMP,
            )}deg`,
          },
        ],
      };
  
      return {
        ...withAnchorPoint(
          transform,
          { x: 0.5, y: 0.5 },
          { width: WIDTH, height: HEIGHT },
        ),
      };
    }, [index]);
  
    const blockStyle = useAnimatedStyle(() => {
      const translateX = interpolate(
        animationValue.value,
        [-1, 0, 1],
        [0, 60, 60],
      );
  
      const translateY = interpolate(
        animationValue.value,
        [-1, 0, 1],
        [0, -40, -40],
      );
  
      const rotateZ = interpolate(
        animationValue.value,
        [-1, 0, 1],
        [0, 0, -25],
      );
  
      return {
        transform: [
          { translateX },
          { translateY },
          { rotateZ: `${rotateZ}deg` },
        ],
      };
    }, [index]);
  
    return (
      <Animated.View
        style={{
          flex: 1,
          justifyContent: "center",
          alignItems: "center",
        }}
        
        accessibilityLabel="answer input"
        accessibilityValue={{ min: 0, max: 100, now: 25, text: '25%' }}
        accessibilityHint={"AGGGG"}
      >
        <Animated.View
          style={[
            {
              backgroundColor: colors[index],
              alignSelf: "center",
              justifyContent: "center",
              alignItems: "center",
              borderRadius: 20,
              width: WIDTH,
              height: HEIGHT,
              shadowColor: "#000",
              shadowOffset: {
                width: 0,
                height: 8,
              },
              shadowOpacity: 0.44,
              shadowRadius: 10.32,
  
              elevation: 16,
            },
            cardStyle,
          ]}
        />
  
        <Animated.Image
          source={fruitItems[index]}
         
          style={[
            {
              width: WIDTH * 0.8,
              borderRadius: 16,
              justifyContent: "center",
              alignItems: "center",
              position: "absolute",
              zIndex: 999,
            },
            blockStyle,
          ]}
          resizeMode={"contain"}
          accessibilityRole="image"
        />
      </Animated.View>
    );
  };

  return (
    <View style={{ flex: 1 }}>
      <Carousel
        {...baseOptions}
        loop
        autoPlay={isAutoPlay}
        withAnimation={{
          type: "spring",
          config: {
            damping: 13,
          },
        }}
        testID={"CAROUSEL_ITEM_0_READY"}
        autoPlayInterval={1500}
        data={props.colors ? props.colors : colors}
        renderItem={({ index, animationValue }) => (
          <Card
            animationValue={animationValue}
            key={index}
            index={index}
          />
        )}
      />
     <Button
        onPress={() => {
          setIsAutoPlay(!isAutoPlay)
        }}
        title={isAutoPlay ? 'Turn off autoplay' : 'Turn on autoplay'}
        color={"red"}
      />
         
    </View>
  );
}



export default AppParellax;
