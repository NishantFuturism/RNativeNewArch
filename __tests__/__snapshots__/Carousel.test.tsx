import React from "react";

import { act, cleanup, fireEvent, render, screen, waitFor } from "@testing-library/react-native";
import AppParellax from "/Users/nishanti/Desktop/RN/futurismts/AppParellax";
import { advanceAnimationByTime, withReanimatedTimer } from "react-native-reanimated/src/reanimated2/jestUtils";
import TestRenderer from 'react-test-renderer';
import Animated from "react-native-reanimated";


  test("renders %s correctly", async () => {
    const colors = ["#1519a1", "#345730", "#34067d"];

    let renderResult: ReturnType<typeof render>;
    withReanimatedTimer(() => {
      // const Carousel = ;
      renderResult = render(<AppParellax  />);
    });
    expect(renderResult!).toBeDefined();
    const { toJSON } = renderResult!;
    withReanimatedTimer(() => {
      act(() => advanceAnimationByTime(2500));
    });
    await waitFor(() => screen.findByTestId("CAROUSEL_ITEM_0_READY"),{timeout : 2000});
    expect(toJSON()).toMatchSnapshot();
  });


// jest.useFakeTimers();

test('stop in the middle of animation', () => {
  const style = { width: 0 };

  const { getByTestId } = render(<AppParellax />);
  let view = getByTestId('CAROUSEL_ITEM_0_READY');
  // const button = getByTestId('button');
  view = view._fiber.stateNode.props;
  expect(view.width).toBe(750);
  expect(view.height).toBe(900);
  expect(view.loop).toBeTruthy();
  expect(view.autoPlay).toBeTruthy();
  expect(view.autoPlayInterval).toBe(1500)

  // fireEvent.press(button);
  withReanimatedTimer(() => {
    advanceAnimationByTime(250); // if whole animation duration is a 500ms
  })
  // style.width = 50; 
  // value of component width after 250ms of animation
  // expect(view).toHaveAnimatedStyle(style);
});


test('render Carousel',async () => {
  // let screen = render(<AppParellax/>)
  // const { toJSON } = screen;
  // const {getAllByLabelText} = screen;
  // console.log(toJSON().children[1].children[0].children[0].props);
  // console.log(toJSON().children[0].props.renderItem);
  // const {debug} = render(<AppParellax/>);
  // debug();

  // const {toJSON} = screen;
  // await waitFor(() => screen.findByAccessibilityValue("AGGGG"),{timeout : 2000});
  // expect(toJSON()).toMatchSnapshot();

  // afterEach(cleanup);

  
  // render(<AppParellax />);


  // const submitButton =  await waitFor(() => screen.findByAccessibilityHint('AGGGG'),{timeout : 2000});

// const submitButton = await screen.getAllBy('AGGGG');
// console.log(submitButton);
// expect(submitButton).toHaveLength(3);


// const element = await screen.getByText('AGGGG');
// console.log(element);


})

test("dummy react test renderer",() => {
//   const testRenderer = TestRenderer.create(<AppParellax />);
// const testInstance = testRenderer.root;

// console.log(testInstance.children);
render(<AppParellax/>);
const element = screen.getByTestId("CAROUSEL_ITEM_0_READY",{
  includeHiddenElements : true
});
expect(element).not.toBeUndefined();
expect(element._fiber.stateNode.props.withAnimation).toStrictEqual({"config": {"damping": 13}, "type": "spring"})
expect(element).toBeDefined();
expect(element).not.toBeFalsy();

// console.log(element);


})