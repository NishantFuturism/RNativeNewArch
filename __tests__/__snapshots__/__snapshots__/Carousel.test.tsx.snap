// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Carousel 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Carousel
    autoPlay={true}
    autoPlayInterval={1500}
    data={
      [
        "#fda282",
        "#fdba4e",
        "#800015",
      ]
    }
    height={900}
    loop={true}
    renderItem={[Function]}
    testID="CAROUSEL_ITEM_0_READY"
    vertical={false}
    width={750}
    withAnimation={
      {
        "config": {
          "damping": 13,
        },
        "type": "spring",
      }
    }
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "red",
            },
          ]
        }
      >
        Turn off autoplay
      </Text>
    </View>
  </View>
</View>
`;

exports[`renders %s correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <Carousel
    autoPlay={true}
    autoPlayInterval={1500}
    data={
      [
        "#fda282",
        "#fdba4e",
        "#800015",
      ]
    }
    height={900}
    loop={true}
    renderItem={[Function]}
    testID="CAROUSEL_ITEM_0_READY"
    vertical={false}
    width={750}
    withAnimation={
      {
        "config": {
          "damping": 13,
        },
        "type": "spring",
      }
    }
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "red",
            },
          ]
        }
      >
        Turn off autoplay
      </Text>
    </View>
  </View>
</View>
`;
